src/main/app/common/country.ts

```diff
@@ -0,0 +1,122 @@
+    this.value = value
+    this.name = name
+  }
+
+  static all (): Country[] {
+    return [
+      Country.WALES,
+      Country.ENGLAND,
+      Country.SCOTLAND,
+      Country.NORTHERN_IRELAND
+    ]
+  }
+
+  static defendantCountries (): Country[] {
+    return [
+      Country.WALES,
+      Country.ENGLAND
+    ]
+  }
+
+  static async isValidClaimantAddress (form: Form<PartyDetails>): Promise<Form<PartyDetails>> {
+    await this.getCountry(form.model.address.postcode).then((country: string) => {
+      if (country.length > 0 && !Country.isClaimantCountry(country)) {
+        let error = new ValidationError()
+        error.property = 'address[postcode]'
+        error.constraints = { amount: 'The country must be England, Wales, Scotland or Northern Ireland' }
+        form.errors.push(new FormValidationError(error))
+      }
+    })
+
+    if (form.model.hasCorrespondenceAddress) {
+      await this.getCountry(form.model.address.postcode).then((country: string) => {
+        if (country.length > 0 && !Country.isClaimantCountry(country)) {
+          let error = new ValidationError()
+          error.property = 'correspondenceAddress[postcode]'
+          error.constraints = { amount: 'The country must be England, Wales, Scotland or Northern Ireland' }
+          form.errors.push(new FormValidationError(error))
+        }
+      })
+    }
+
+    return form
+  }
+
+  static async isValidDefendantAddress (form: Form<PartyDetails>): Promise<Form<PartyDetails>> {
+    await this.getCountry(form.model.address.postcode).then((country: string) => {
+      if (country.length > 0 && !Country.isDefendantCountry(country)) {
+        let error = new ValidationError()
+        error.property = 'address[postcode]'
+        error.constraints = { amount: 'The country must be England or Wales' }
+        form.errors.push(new FormValidationError(error))
+      }
+    })
+
+    if (form.model.hasCorrespondenceAddress) {
+      await this.getCountry(form.model.address.postcode).then((country: string) => {
+        if (country.length > 0 && !Country.isDefendantCountry(country)) {
+          let error = new ValidationError()
+          error.property = 'correspondenceAddress[postcode]'
+          error.constraints = { amount: 'The country must be England or Wales' }
+          form.errors.push(new FormValidationError(error))
+        }
+      })
+    }
+
+    return form
+  }
+
+  static isClaimantCountry (value: string): boolean {
+    return Country.all().filter(country => country.name.toLowerCase() === value.toLowerCase()).length > 0
+  }
+
+  static isDefendantCountry (value: string): boolean {
+    return Country.defendantCountries().filter(country => country.name.toLowerCase() === value.toLowerCase()).length > 0
+  }
+
+  static async getCountry (postcode: string): Promise<string> {
+    if (postcode === undefined || postcode.length === 0) {
+      return ''
+    } else {
+      let country = ''
+      await postcodeClient.lookupPostcode(postcode)
```

Comment:
why are you looking this up again? isn't it already in the model?

Reply:
The model doesn't save the returned object from the javascript call and it is possible that someone manually enters the address. we still need to do the check on those entries

Reply:
Please do not use `then / catch` together with async / await.